src/v2/coins/abstractUtxoCoin.ts

```diff
@@ -803,50 +796,83 @@ class AbstractUtxoCoin extends BaseCoin {
+      const publicKeys = [publicKey];
+      const signatures = [signature];
+      const pubScript = bitcoin.script.pubKeyHash.output.encode(bitcoin.crypto.hash160(publicKey));
+
+      return { isSegwitInput, inputClassification, signatures, publicKeys, pubScript };
+    } else if (inputClassification === bitcoin.script.types.P2SH
+        || inputClassification === bitcoin.script.types.P2WSH) {
+      // Note the assumption here that if we have a p2sh or p2wsh input it will be multisig (appropriate because the
+      // BitGo platform only supports multisig within these types of inputs). Signatures are all but the last entry in
+      // the decompiledSigScript. The redeemScript/witnessScript (depending on which type of input this is) is the last
+      // entry in the decompiledSigScript (denoted here as the pubScript). The public keys are the second through
+      // antepenultimate entries in the decompiledPubScript. See below for a visual representation of the typical 2-of-3
+      // multisig setup:
+      //
+      // decompiledSigScript = 0 <sig1> <sig2> <pubScript>
+      // decompiledPubScript = 2 <pub1> <pub2> <pub3> 3 OP_CHECKMULTISIG
+      const signatures = decompiledSigScript.slice(0, -1);
+      const pubScript = _.last(decompiledSigScript);
+      const decompiledPubScript = bitcoin.script.decompile(pubScript);
+      const publicKeys = decompiledPubScript.slice(1, -2);
+
+      // Op codes 81 through 96 represent numbers 1 through 16 (see https://en.bitcoin.it/wiki/Script#Opcodes), which is
+      // why we subtract by 80 to get the number of signatures (n) and the number of public keys (m) in an n-of-m setup.
+      const len = decompiledPubScript.length;
+      const nSignatures = decompiledPubScript[0] - 80;
+      const nPubKeys = decompiledPubScript[len - 2] - 80;
+
+      // Due to a bug in the implementation of multisignature in the bitcoin protocol, a 0 is added to the signature
+      // script, so we add 1 when asserting the number of signatures matches the number of signatures expected by the
+      // pub script. Also, note that we consider a signature script with the the same number of signatures as public
+      // keys (+1 as noted above) valid because we use placeholder signatures when parsing a half-signed signature
+      // script.
+      if (signatures.length !== nSignatures + 1 && signatures.length !== nPubKeys + 1) {
+        throw new Error(`expected ${nSignatures} or ${nPubKeys} signatures, got ${signatures.length - 1}`);
+      }
+
+      if (publicKeys.length !== nPubKeys) {
+        throw new Error(`expected ${nPubKeys} public keys, got ${publicKeys.length}`);
+      }
 
-        const pubScript = bitcoin.script.pubKeyHash.output.encode(bitcoin.crypto.hash160(publicKey));
-        return { signatures, publicKeys, isSegwitInput, inputClassification, pubScript: pubScript };
+      const lastOpCode = decompiledPubScript[len - 1];
+      if (lastOpCode !== bitcoin.opcodes.OP_CHECKMULTISIG) {
```

Comment:
actually I think we can use the bitcoin lib stuff to replace most of this here:

```
          const { redeemScript, redeemScriptSig } = utxolib.script.scriptHash.input.decode(input.script);
          const { m, pubKeys } = utxolib.script.multisig.output.decode(redeemScript);
          const multiSigScript = utxolib.script.multisig.input.decode(redeemScriptSig);
          // throw if `m !== multiSigScript.length`
```
(with `utxolib=bitcoin`)

Reply:
While I do prefer this as it would make our code a lot more readable, as it currently stands this would not work. `scriptHash.input.decode()` does not support half-signed multisig scripts. However, `parseSignatureScript` needs to be able to handle half-signed multisig scripts. The change in `bitgo-utxo-lib` needed to make `scriptHash.input.decode()` support half-signed multisig scripts would be small but I am not sure if that function intentionally did not handle those types of scripts. I would suggest leaving this code as is and discussing whether or not `scriptHash.input.decode()` should handle half-signed multisig scripts outside the scope of this PR.